File: problem_031.c
// -*- compile-command: "gcc -o problem_031 problem_031.c -Wall -O3" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>

int addCounter(int *counter, int start, int stop)
{
    int sum = 0;
    int i;

    for (i=start; i<=stop; i++)
        sum += counter[i];

    return sum;
}

int main()
{
    printf("Project Euler - Problem 30:\n"
           "How many different ways can 2 pounds be made using any number of coins?\n\n");

    int counter[8] = {0,0,0,0,0,0,0,0};
    int count = 0;

    while (counter[6] <= 200)
    {
        counter[0]=counter[0]+1;
        if (counter[0] > 200 - addCounter(counter, 1,7))
        {
            counter[0] = 0;
            counter[1] += 2;
        }
        if (counter[1] > 200 - addCounter(counter, 2,7))
        {
            counter[1] = 0;
            counter[2] += 10;
        }
        if (counter[2] > 200 - addCounter(counter, 3,7))
        {
            counter[2] = 0;
            counter[3] += 5;
        }
        if (counter[3] > 200 - addCounter(counter, 4,7))
        {
            counter[3] = 0;
            counter[4] += 20;
        }
        if (counter[4] > 200 - addCounter(counter, 5,7))
        {
            counter[4] = 0;
            counter[5] += 50;
        }
        if (counter[5] > 200 - addCounter(counter, 6,7))
        {
            counter[5] = 0;
            counter[6] += 100;
        }
        if (counter[6] > 200 - counter[7])
        {
            counter[6] = 0;
            counter[7] += 200;
        }
        if ( addCounter(counter, 0,7) == 200)
        {
            count++;
            if ( counter[7] == 200 ) break;
        }
    }

    printf("Ways: %d\n", count);

    return 0;
}

File: problem_031.js
#!/usr/bin/node

console.log("Project Euler - Problem 31:");
console.log("How many different ways can 2 pounds be made using any number of coins?\n");

var ways = 0;

// This is the naive approach, but it's fun
for (var twohundred=0; twohundred<=200; twohundred+=200){
    for (var onehundred=0; onehundred<=200; onehundred+=100){
        if (twohundred + onehundred > 200) break;
        for (var fifty=0; fifty<=200; fifty+=50){
            if (twohundred + onehundred + fifty > 200) break;
            for (var twenty=0; twenty<=200; twenty+=20){
                if (twohundred + onehundred + fifty + twenty > 200) break;
                for (var ten=0; ten<=200; ten+=10){
                    if (twohundred + onehundred + fifty + twenty + ten > 200) break;
                    for (var five=0; five<=200; five+=5){
                        if (twohundred + onehundred + fifty + twenty + ten + five > 200) break;
                        for (var two=0; two<=200; two+=2){
                            if (twohundred + onehundred + fifty + twenty + ten + five + two > 200) break;
                            ways += 1; // ones can fill in the remainder if the sum is <= 200
                        }
                    }
                }
            }
        }
    }
}

console.log(ways);
